src/StackExchange.Redis/PhysicalConnection.cs

```diff
@@ -1632,9 +1632,23 @@ static bool TryGetPubSubPayload(in RawResult value, out RedisValue parsed, bool
         internal void GetHeadMessages(out Message? now, out Message? next)
         {
             now = _activeMessage;
-            lock(_writtenAwaitingResponse)
+            bool haveLock = false;
+            try
+            {
+                // careful locking here; a: don't try too hard (this is error info only), b: avoid deadlock (see #2376)
+                Monitor.TryEnter(_writtenAwaitingResponse, 10, ref haveLock);
+                if (haveLock)
+                {
+                    _writtenAwaitingResponse.TryPeek(out next);
+                }
+                else
+                {
+                    next = null;
```

Comment:
Should we indicate "unknown" here if we couldn't get the lock? It's likely there was _something_ there, but faking a message just to say that in this structure seems odd too - thoughts?

Reply:
@NickCraver  try now

Reply:
LGTM, thanks! That'll make if we see it a bit easier to reason <3